Support %YAML 1.2 directives

perlpunk opened the pull request:
This will allow `%YAML 1.2` directives.

See also #20

No changes are needed regarding tag directives.

In YAML 1.2 tag directives are for the following document only.
This is already implemented like that in libyaml.

We would rather have to fix the code if we want to have the correct behaviour
(global directives) in YAML 1.1. This would be a bit more complicated, as we
would have to save the default version and the current version in the parser
object.

New passing parser tests:

* 27NA: Spec Example 5.9. Directive Indicator
* 6ZKB: Spec Example 9.6. Stream
* 9DXL: Spec Example 9.6. Stream [1.3]
* RTP8: Spec Example 9.2. Document Markers

New failing error parser tests (before they were errors for the wrong reason):

* EB22: Missing document-end marker before directive
* RHX7: YAML directive without document end marker

Tests are currectly failing because I cannot edit the skiplist from this PR.
Not sure why travis doesn't show up here:
https://travis-ci.org/github/yaml/libyaml/builds/668849280

ingydotnet commented:
I'm on the fence on this one. On one hand it's a simple fix. One the other hand I want to see directives go away in YAML. So this would encourage using a broken (imho) YAML language feature in libyaml. That would make later changes to the language harder to put in place.

I'd like to see how the RFC process directives goes, before making a decision on this.

NOTE: I don't want to hash out the issues around directives here. Let's do that in the RFC process as it emerges.

perlpunk commented:
YAML 1.2 will be around for a long while even after 1.3 is created.
The YAML directive is in the 1.1 spec and in the 1.2 spec.

Allowing this will be very useful for systems currently having to deal with 1.1 files and wanting to upgrade to 1.2.

I already talked to ansible people a while ago, and they seemed to see this as a good solution (using explicit %YAML 1.2 directives where documents should be interpreted like 1.2 regarding the schema, and defaulting to 1.1).

I have no idea why you would block this because of 1.3. People need this feature. I'm a YAML user, I'm also using ansible, and colleagues are using saltstack.

ingydotnet commented:
As Larry says, TMTOWTDI.

A YAML document does not stand alone. It requires a context. It needs more context than `%YAML` or `%TAG` directives can provide. If we come up with better ways to provide these contexts declaratively, it is possible that we can get rid of YAML directives altogether, and simplify the language.

I also checked with some ansible people and looked at the ansible documentation. Version declaration is not used much if at all. Looking a saltstack, I come to the same conclusion.

This likely has to do with the fact that there are not many implementation choices between 1.1 and 1.2.

If we make libyaml 1.2 compliant, there might be more of a need. But there is more than one way to declare the yaml version being used, just as there is more than one way to declare the other context like tags and schemas.

If yaml version declaration usage is not popular now, and there are better ways to do it, and we intend to move in that direction, I don't want to encourage people to start using it. That is the point I was trying to make in my last comment.

I don't think it's fair to say "People need this feature"; especially when it is barely in current use. That is a discussion to be had in the language design process.

perlpunk commented:
Added/removed tests in `tests/run-test-suite/test/blacklist/libyaml-parser*`. Committed. Pushed.
Wrote down commit sha.
Rebased this PR to `release/0.2.3`, wrote down commit sha.
Added line in `tests/run-test-suite/conf/pin.tsv` with the new commit shas.
Comitted, pushed.
Now merging this PR and hoping that tests will use the correct data and pass.

perlpunk commented:
Oh, of course it failed, because I used "squash commits" and it added the PR number to the commit, so the SHA is different. *cries*
